### PR TITLE
Pretty Reporting

### PR DESCRIPTION
**Pull request by wdahlenburg**
Description
---------------

~~This PR is a work in progress (WIP).~~

This is to address #762. This ended up rewriting/clobbering more code than I was hoping for. There are definitely bugs, so please address them and I'll work to balance the new behavior vs the old.

The major upgrade of this PR is that reports will return a combined format in a single output for all urls. The reports were originally written to write one report per url, which is fine, but batch reports end up looking ugly and contain redundant metadata. This required a heavy rewrite to the report_manager.py script. 

The current PR focuses on writing a single report. The previous code performed some auto-saving and appending. That has been removed in an effort to get this working. Previously if you added the --simple-report or similar flags you would see the standard log file written along with selected output file. The log + report behavior has also been removed, but would be pretty simple to add back if desired.

The base_report.py script still needs some work due to some of these pending questions. Each report was modified to account for the url iteration. I used a bit of #761, but I'd like to merge that PR and resolve any conflicts between the styles. 

A major simplification of reporting was made by dropping the --<format>-report flag. The outputFile (-o) and outputFormat (--format) flags were added, which clean up some of the long if-then statements.  

I'll add comparisons in the comments of the PR of report output using batch mode to demonstrate the changes.

---
## Update 4/21/21

This PR should be good to go for testing.

### Usage

```
# Basic example (single url)
$ python3 ./dirsearch.py -u https://example.com/ --format json -o example_dirsearch.json

# Basic example (multiple urls)
$ python3 ./dirsearch.py -l urls.txt --format json -o batch_report.json
```

The same functionality should be provided with the auto-save functionality. You should be able to leave out the ```-o``` flag and just specify the output format to use the standard report location. You should be able to run a ```tail -f <output_file>``` while dirsearch is running and watch any changes to the output file.

### CODE REVIEW

**Comment by wdahlenburg** [comment]
Dirsearch on ae36ca7

```bash
$ python3 ./dirsearch.py -l /tmp/urls.txt -x 403,500-599 -F -w /tmp/list.txt --json-report /tmp/json-report.json
```

```json
{
    "https://example.com:443/": [
        {
            "content-length": 37261,
            "path": "/.m/",
            "redirect": null,
            "status": 200
        }
    ],
    "time": "Mon Mar  8 21:59:01 2021"
}{
    "https://example.com:443/": [
        {
            "content-length": 37261,
            "path": "/.m/",
            "redirect": null,
            "status": 200
        },
        {
            "content-length": 37261,
            "path": "/.meta/",
            "redirect": null,
            "status": 200
        }
    ],
    "time": "Mon Mar  8 21:59:01 2021"
}{
    "http://127.0.0.1:8090/": [
        {
            "content-length": 16,
            "path": "/index.html",
            "redirect": null,
            "status": 200
        }
    ],
    "time": "Mon Mar  8 21:59:01 2021"
}
```

WIP PR

```bash
$ python3 ./dirsearch.py -l /tmp/urls.txt -x 403,500-599 -F -w /tmp/list.txt -o /tmp/json-report2.json --format json
```

```json
{
    "info": {
        "time": "Mon Mar  8 20:37:46 2021"
    },
    "results": [
        {
            "https://example.com:443/": [
                {
                    "content-length": 37261,
                    "path": "/.meta/",
                    "redirect": null,
                    "status": 200
                },
                {
                    "content-length": 37261,
                    "path": "/.m/",
                    "redirect": null,
                    "status": 200
                }
            ]
        },
        {
            "http://127.0.0.1:8090/": [
                {
                    "content-length": 16,
                    "path": "/index.html",
                    "redirect": null,
                    "status": 200
                }
            ]
        }
    ]
}
```

**Comment by wdahlenburg** [comment]
Dirsearch on ae36ca7

```
$ python3 ./dirsearch.py -l /tmp/urls.txt -x 403,500-599 -F -w /tmp/list.txt --xml-report /tmp/xml-report.xml
```

```xml
<?xml version="1.0"?>
<time>Mon Mar  8 22:05:47 2021</time>
<target url="https://example.com:443/">
 <info path="/.m/">
  <status>200</status>
  <contentLength>37261</contentLength>
  <redirect>None</redirect>
 </info>
</target>
<?xml version="1.0"?>
<time>Mon Mar  8 22:05:47 2021</time>
<target url="https://example.com:443/">
 <info path="/.m/">
  <status>200</status>
  <contentLength>37261</contentLength>
  <redirect>None</redirect>
 </info>
 <info path="/.meta/">
  <status>200</status>
  <contentLength>37261</contentLength>
  <redirect>None</redirect>
 </info>
</target>
<?xml version="1.0"?>
<time>Mon Mar  8 22:05:47 2021</time>
<target url="http://127.0.0.1:8090/">
 <info path="/index.html">
  <status>200</status>
  <contentLength>16</contentLength>
  <redirect>None</redirect>
 </info>
</target>
```

WIP PR

```
$ python3 ./dirsearch.py -l /tmp/urls.txt -x 403,500-599 -F -w /tmp/list.txt -o /tmp/xml-report2.xml --format xml
```

```xml
<?xml version="1.0"?>
<dirsearchScan args="./dirsearch.py -l /tmp/urls.txt -x 403,500-599 -F -w /tmp/list.txt -o /tmp/xml-report2.xml --format xml" time="Mon Mar  8 22:09:59 2021">
 <target url="https://example.com:443/">
  <info path="/.m/">
   <status>200</status>
   <contentLength>37261</contentLength>
   <redirect></redirect>
  </info>
  <info path="/.meta/">
   <status>200</status>
   <contentLength>37261</contentLength>
   <redirect></redirect>
  </info>
 </target>
 <target url="http://127.0.0.1:8090/">
  <info path="/index.html">
   <status>200</status>
   <contentLength>16</contentLength>
   <redirect></redirect>
  </info>
 </target>
</dirsearchScan>
```

**Comment by maurosoria** [comment]
Hello @wdahlenburg, how you doing?

I am trying to figure out why are you removing the Tailable Report (very useful by the way).

**Comment by wdahlenburg** [comment]
Hey @maurosoria. I'm doing pretty well. How about yourself? 

Being a work-in-progress, I mainly wanted to get the reports restructured to support results from multiple hosts. In the short-term my concern around tailable reports was that structured files (JSON and XML) would not stay structured. I think it should be possible to add back the tailable support by updating the output once every host is completed. The unstructured formats should be able to write as soon as results are captured, but I'll have to see how that looks in the code.

I need to catch up to the master branch and see what I can do.

**Comment by wdahlenburg** [comment]
The latest version should clean up some more code and allow for file writes after results are available for each host. The json and xml formats will overwrite the entire file every time to guarantee the structured format. The one minor thing that occurs is that the timestamp is overwritten on each update. This could be fixed to the first write time or some other value like scan start time. 

The other formats will track which hosts have already been written to and skip over them on subsequent file saves. This appears to balance tailable reports while maintaining the structure of the json/xml. Try a ```tail -f <file>``` on any formatted batch output file and let me know if this works. 

The report manager now maintains a reference to the report type which allows for the entries to be maintained as new data is present.

**Comment by maurosoria** [comment]
Hello @wdahlenburg ! Thanks for contributing to dirsearch project!

This PR looks good but has to be tested and review. 

Please wait until the process is completed.

Regards,
Mauro

**Comment by shelld3v** [comment]
Hi @wdahlenburg, the output looks weird. Is this what you expected too?

```

  _|. _ _  _  _  _ _|_    v0.4.1
 (_||| _) (/_(_|| (_| )

Extensions: php, aspx, jsp, html, js | HTTP method: GET | Threads: 30 | Wordlist size: 10861

Error Log: /home/admin/dirsearch/logs/errors-21-04-21_08-56-59.log
Output File: /home/admin/dirsearch/reports/example.com/_21-04-21_08-56-59.txt


Target: http://example.com/

[08:57:00] Starting:
```

**Comment by wdahlenburg** [comment]
@shelld3v Yeah I agree the spacing between the Error Log, Output File, and Target is off. 

This is what I have so far in terms of bugs/improvements. Keep adding comments with any issues. 

- [ ] A bug exists when a url with a folder is specified (https://example.com/foo/) along with a format type, but no output file selected. The url path isn't get converted into a writable file path.
    ```python3 ./dirsearch.py -u https://example.com/foo/ --format json``` -> ```FileNotFoundError: [Errno 2] No such file or directory: ...```
- [ ] Fix spacing in headers between logs and target
- [ ] Add a check to see if the format type is plain or simple and set the extension to .txt when no output file is specified, but a format is.

**Comment by wdahlenburg** [comment]
All three of these have been fixed. 

I personally find the "AutoSave path" to be redundant information, since the batch report will contain the same folder path. It can be left as-is if it's helpful or removed. 

I'm going to remove the [WIP] label as this should be ready for further testing.

**Comment by shelld3v** [comment]
Hi, great work! Does this also fix #219?

**Comment by shelld3v** [comment]
```
Extensions: php, aspx, jsp, html, js | HTTP method: GET | Threads: 30 | Wordlist size: 10861

Error Log: /home/admin/dirsearch/logs/errors-21-04-22_21-21-04.log
Output File: /home/admin/dirsearch/reports/example.com/_21-04-22_21-21-09.txt

Target: http://example.com/

[21:21:10] Starting:
```

This is intended? If that the case, should there be a new line (`\n`) between them (`Error Log` and `Output File`)? And I still more like the old output

**Comment by shelld3v** [comment]
```
Extensions: php, aspx, jsp, html, js | HTTP method: GET | Threads: 30 | Wordlist size: 10861

Error Log: /home/admin/dirsearch/logs/errors-21-04-22_21-24-59.log
AutoSave path: /home/admin/dirsearch/reports/BATCH-21-04-22_21-24-59
Output File: /home/admin/dirsearch/reports/BATCH-21-04-22_21-24-59/BATCH.txt

Target: http://example.com/
...
```

Not nice!

**Comment by wdahlenburg** [comment]
> Hi, great work! Does this also fix #219?

Yes. All hosts should be reported in any output file once the host has completed.

**Comment by shelld3v** [comment]
I also think this looks better (I believe I have OCD):

```
  Reports:
    -o FILE              Output file
    --format=FORMAT      Report format (Available: simple, plain, json, xml, md, csv)
```

**Comment by wdahlenburg** [comment]
> Oh wait, there is a problem! If I stop in the middle of the scan, the output file will be empty (I tried with `CTRL+C -> e`). Can you please fix this?

Right now the output file is written/appended to after each host is completed. I can add a one-liner to fix this situation, where a `CTRL+C -> e` will write the latest results to the output file and then exit. Would this work?

**Comment by shelld3v** [comment]
> > Oh wait, there is a problem! If I stop in the middle of the scan, the output file will be empty (I tried with `CTRL+C -> e`). Can you please fix this?
> 
> Right now the output file is written/appended to after each host is completed. I can add a one-liner to fix this situation, where a `CTRL+C -> e` will write the latest results to the output file and then exit. Would this work?

I don't think so. What happens if the computer has an error then suddenly freezes/shutdown/closes tasks? The user will lose all results from the scan (a really really long scan)?

**Comment by wdahlenburg** [comment]
> > > Oh wait, there is a problem! If I stop in the middle of the scan, the output file will be empty (I tried with `CTRL+C -> e`). Can you please fix this?
> > 
> > 
> > Right now the output file is written/appended to after each host is completed. I can add a one-liner to fix this situation, where a `CTRL+C -> e` will write the latest results to the output file and then exit. Would this work?
> 
> I don't think so. What happens if the computer has an error then suddenly freezes/shutdown/closes tasks? The user will lose all results from the scan (a really really long scan)?

I think they should lose the results from that host, rather than believe they fully completed the wordlist against that host. 

Take the following list of urls:
```
https://example.com/
https://foobar.com/
https://foo.com/
```

```
https://example.com/     -> Host complete
https://foobar.com/      -> Scan fails here
https://foo.com/
```
In this scenario only the results for https://example.com would be written to the output file. This would allow a user of dirsearch to recognize that https://foobar.com/ did not fully complete. This could allow for some type of resume feature similar to nmap.

**Comment by shelld3v** [comment]
But the fact that a lot of scans require even an hour or two hours to scan, losing the results because of an outage is really bad. The reason a scan takes so much time is because: the server response speed is slow or we need to run with fewer threads to prevent from noticing the firewall, the wordlist is very huge and finally, recursive scans take forever

**Comment by shelld3v** [comment]
@wdahlenburg What is your final decision? I likely want to save entry by entry, because I usually run dirsearch at night, and when I'm too tired to wait for the scan to finish, I just stop. But it's ok if you want to keep your idea, I will look for another solution, maybe sessions saving.

**Comment by wdahlenburg** [comment]
I fixed up the help statement to match @shelld3v 's comments and also force a save on ```CTRL+C -> e```. I'll come back to the header format since that still needs to be fixed. I'm assuming there should be a line in between each output file. 

I want to add a small validation line to verify that the selected format is a valid choice. 

@maurosoria What's your opinion on shelldev's recent comments? dee2d5c is the latest commit. I can write the code to save on every new result or only after a host is completed. I'm good with it either way.

**Comment by maurosoria** [comment]
@wdahlenburg check your email

**Comment by shelld3v** [comment]
@wdahlenburg https://join.slack.com/t/dirsearch/shared_invite/zt-pe9cq4md-7WbbZtxRWyUkeaW1zw7KrA

**Comment by shelld3v** [comment]
@wdahlenburg Also, have u thought about printing the full path of the output file? (Not only `arguments.outputFile`). I believe it's important because now dirsearch can be run anywhere (with `setup.py`).

>I'm assuming there should be a line in between each output file.

Nah, it's ok now

**Comment by shelld3v** [comment]
>I think they should lose the results from that host, rather than believe they fully completed the wordlist against that host.

You can write `"state": "Uncompleted"`

**Comment by wdahlenburg** [comment]
I fixed up the help statement to match @shelld3v 's comments. 

I want to add a small validation line to verify that the selected format is a valid choice. 

@maurosoria What's your opinion on this? dee2d5c is the latest commit. I can write the code to save on every new result or only after a host is completed. 

> @wdahlenburg Also, have u thought about printing the full path of the output file? (Not only `arguments.outputFile`). I believe it's important because now dirsearch can be run anywhere (with `setup.py`).
> 
> > I'm assuming there should be a line in between each output file.
> 
> Nah, it's ok now

Good call. I have just been testing with a full path. This will be good to add

**Comment by wdahlenburg** [comment]
> @wdahlenburg What is your final decision? I likely want to save entry by entry, because I usually run dirsearch at night, and when I'm too tired to wait for the scan to finish, I just stop. But it's ok if you want to keep your idea, I will look for another solution, maybe sessions saving.

@shelldev can you please test out the [reporting_dev_write_on_result](https://github.com/wdahlenburg/dirsearch/tree/reporting_dev_write_on_result) branch on my profile. It's based off this branch, but it should perform a write operation every time there is a result. Surprisingly enough the markdown format gave me the hardest time. 

I actually ended up including a check to see if the host had been completed, so this could be used to indicate in the structured output files (json and xml) that the host hasn't been completed. 

Let me know your thoughts and I can merge this in to this PR branch.

**Comment by shelld3v** [comment]
>@shelldev can you please test out the reporting_dev_write_on_result branch on my profile. It's based off this branch, but it >should perform a write operation every time there is a result. Surprisingly enough the markdown format gave me the hardest >time.
>
>I actually ended up including a check to see if the host had been completed, so this could be used to indicate in the structured >output files (json and xml) that the host hasn't been completed.
>
>Let me know your thoughts and I can merge this in to this PR branch.

```
Extensions: php, aspx, jsp, html, js | HTTP method: GET | Threads: 30 | Wordlist size: 10861
Output File: tmp

Error Log: /home/admin/dirsearch/logs/errors-21-04-23_21-45-41.log
```
Fix this! And, that is really impressive, I love it. But as I mentioned:

>You can write `"state": "Uncompleted",`

**Comment by wdahlenburg** [comment]
The write functionality was switched to write on every result instead of only after each host was complete. 

The headers were fixed for batch and non-batch mode. There shouldn't be missing newlines anymore.

The absolute path will display as the output file when a user provides a relative path to the -o argument. 

A small check was added to ensure that the passed format is a valid type. This prevents someone from specifying --format blah and the output file would end in ".blah". 

I'm going to put a hold on including the completion status for now.

**Comment by shelld3v** [comment]
Hi @wdahlenburg,

This is perfect, but I have a question: If all scans' results are saved into only a single file, then why we should create a whole new folder for it? The `TARGETS.txt` file is useless because we can find targets in `BATCH.txt`

Also:)

```
AutoSave path: /home/admin/dirsearch/reports/BATCH-21-04-25_14-35-50

Output File: /home/admin/dirsearch/reports/BATCH-21-04-25_14-35-50/BATCH.txt

Error Log: /home/admin/dirsearch/logs/errors-21-04-25_14-35-50.log

Target: http://testphp.vulnweb.com/

[14:35:51] Starting:
[14:35:56] 301 -  169B  - /.idea  ->  http://testphp.vulnweb.com/.idea/
[14:35:56] 200 -  951B  - /.idea/
[14:35:56] 200 -    6B  - /.idea/.name
[14:35:56] 200 -  171B  - /.idea/encodings.xml
CTRL+C detected: Pausing threads, please wait...
[e]xit / [c]ontinue / [s]kip target: s
-------------------WHERE IS THE NEW LINE HERE?------------------------
Target: http://example.com/

[14:36:00] Starting:
[14:36:03] 403 -  345B  - /.dep.inc
[14:36:03] 403 -  345B  - /.env~
[14:36:04] 403 -  345B  - /.gitignore~
CTRL+C detected: Pausing threads, please wait...
```

**Comment by wdahlenburg** [comment]
> Hi @wdahlenburg,
> 
> This is perfect, but I have a question: If all scans' results are saved into only a single file, then why we should create a whole new folder for it? The `TARGETS.txt` file is useless because we can find targets in `BATCH.txt`
> 
> Also:)
> 
> ```
> AutoSave path: /home/admin/dirsearch/reports/BATCH-21-04-25_14-35-50
> 
> Output File: /home/admin/dirsearch/reports/BATCH-21-04-25_14-35-50/BATCH.txt
> 
> Error Log: /home/admin/dirsearch/logs/errors-21-04-25_14-35-50.log
> 
> Target: http://testphp.vulnweb.com/
> 
> [14:35:51] Starting:
> [14:35:56] 301 -  169B  - /.idea  ->  http://testphp.vulnweb.com/.idea/
> [14:35:56] 200 -  951B  - /.idea/
> [14:35:56] 200 -    6B  - /.idea/.name
> [14:35:56] 200 -  171B  - /.idea/encodings.xml
> CTRL+C detected: Pausing threads, please wait...
> [e]xit / [c]ontinue / [s]kip target: s
> -------------------WHERE IS THE NEW LINE HERE?------------------------
> Target: http://example.com/
> 
> [14:36:00] Starting:
> [14:36:03] 403 -  345B  - /.dep.inc
> [14:36:03] 403 -  345B  - /.env~
> [14:36:04] 403 -  345B  - /.gitignore~
> CTRL+C detected: Pausing threads, please wait...
> ```

Good question. I'd be fine removing the auto save path. It does store the list of targets, which could potentially be helpful if you are using the `--stdin` mode. 

Yes I agree that there should be a newline included when a host is skipped. Would it make sense to consider a skipped host as completed?

**Comment by wdahlenburg** [comment]
@shelld3v The newline was fixed whenever you skip a host. I'm going to consider a skipped host to be completed. This helps ensure the spacing is constant on the markdown report whenever a host is skipped. 

I pulled out the auto-save functionality because the reports will do this automatically. The CLI args will override whatever exists in the default.conf file. This file was modified to mirror the CLI changes.

**Review by shelld3v — APPROVED** [review]

File: lib/controller/controller.py [hunk 1]
```diff
@@ -368,29 +363,38 @@ def setupBatchReports(self):
...
+            if self.arguments.autoSaveFormat == 'plain' or self.arguments.autoSaveFormat == 'simple':
+                return ".txt"
+            else:
+                return ".{0}".format(self.arguments.autoSaveFormat)
+        else:
+            return ".txt"
 
+    def setupReports(self):
+        if self.arguments.outputFile is not None:
+            self.output.outputFile(self.arguments.outputFile)
```
[the remark this answers is not in the document]
Comment (wdahlenburg) [reply lib/controller/controller.py]: The absolute path is set in https://github.com/maurosoria/dirsearch/blob/2d1c7f39a4b4bb7a52e61bd0ef15d24794367d8d/lib/controller/controller.py#L364. 

A helper method was added to the FileUtils object: https://github.com/maurosoria/dirsearch/blob/2d1c7f39a4b4bb7a52e61bd0ef15d24794367d8d/lib/utils/file_utils.py#L90

File: lib/core/argument_parser.py [hunk 2]
```diff
@@ -602,12 +598,9 @@ def parseArguments(self):
...
         # Report Settings
         reports = OptionGroup(parser, "Reports")
-        reports.add_option("--simple-report", action="store", dest="simpleOutputFile", default=None, metavar="OUTPUTFILE")
-        reports.add_option("--plain-text-report", action="store", dest="plainTextOutputFile", default=None, metavar="OUTPUTFILE")
-        reports.add_option("--json-report", action="store", dest="jsonOutputFile", default=None, metavar="OUTPUTFILE")
-        reports.add_option("--xml-report", action="store", dest="xmlOutputFile", default=None, metavar="OUTPUTFILE")
-        reports.add_option("--markdown-report", action="store", dest="markdownOutputFile", default=None, metavar="OUTPUTFILE")
-        reports.add_option("--csv-report", action="store", dest="csvOutputFile", default=None, metavar="OUTPUTFILE")
+        reports.add_option("-o", action="store", dest="outputFile", default=None, metavar="OUTPUTFILE")
+        reports.add_option("--format", action="store", dest="outputFormat", default=None, metavar="simple",
```
[the remark this answers is not in the document]
Comment (wdahlenburg) [reply lib/core/argument_parser.py]: You should be able to specify both the output file and format in the default.conf file. Whatever is specified in there will be selected unless CLI arguments are given, which will override what was set in the default.conf file. 

If a user happens to comment out both entries in the default.conf file and provided no CLI args, then the default plain report is used. The output file will either be the default BATCH.txt or regular host_date syntax depending on if it's batch or not. https://github.com/maurosoria/dirsearch/blob/2d1c7f39a4b4bb7a52e61bd0ef15d24794367d8d/lib/controller/controller.py#L404

**Comment by shelld3v** [comment]
@maurosoria I am going to merge this now!

**Comment by wdahlenburg** [comment]
I forgot to update the Readme.md file with the help banner updates. New PR incoming

**Pull request merged**